mqueue.c

```diff
@@ -44,42 +63,78 @@ int MUTEX_LOCK(struct mq_hdr* hdr)
     }
 }
 
+/* Returns 0 on success */
 int COND_WAIT(struct mq_hdr* hdr)
 {
+	DWORD dwWaitResult;
+
 	ReleaseMutex(hdr->mqh_lock);
-	WaitForSingleObject(hdr->mqh_wait, INFINITE);
+	dwWaitResult = WaitForSingleObject(hdr->mqh_wait, INFINITE);
+	if(dwWaitResult != WAIT_OBJECT_0) {
+		return EINVAL;
+	}
+
 	return MUTEX_LOCK(hdr);
 }
+
+/* Returns 0 on success */
 int COND_TIMED_WAIT(struct mq_hdr* hdr, const struct timespec* abstime)
 {
 	DWORD dwWaitResult;
 
 	ReleaseMutex(hdr->mqh_lock);
 	dwWaitResult = WaitForSingleObject(hdr->mqh_wait, (DWORD) abstime->tv_sec);
-	WaitForSingleObject(hdr->mqh_lock,(DWORD) abstime->tv_sec);
- 
-    return dwWaitResult == WAIT_OBJECT_0;
+	switch(dwWaitResult) {
+	    case WAIT_OBJECT_0:
+	        break
+	    case WAIT_TIMEOUT:
+	        return ETIMEDOUT;
+	    default:
+	        return EINVAL;
+	}
+
+	dwWaitResult = WaitForSingleObject(hdr->mqh_lock,(DWORD) abstime->tv_sec);
+	if(dwWaitResult == WAIT_OBJECT_0) {
+	    return 0;
+	}
+
+    if(dwWaitResult == WAIT_TIMEOUT) {
+        return ETIMEDOUT;
+    }
+
+    return EINVAL;
 }
+
+/* Returns 0 on success */
 int COND_SIGNAL(struct mq_hdr* hdr)
 {
-	SetEvent(hdr->mqh_wait);
-	return 0;
+	BOOL result = SetEvent(hdr->mqh_wait);
```

Comment:
SetEvent success is non-zero, however pthread_cond_signal is 0 on success, it should match pthread_cond_signal.  Of Interested is that the return value is currently not checked